test/hotspot/gtest/runtime/test_os.cpp

```diff
@@ -976,3 +979,34 @@ TEST_VM(os, vm_min_address) {
 #endif
 }
 
+#if !defined(_WINDOWS) && !defined(_AIX)
+TEST_VM(os, free_without_uncommit) {
+  const size_t page_sz = os::vm_page_size();
+  const size_t pages = 64;
+  const size_t size = pages * page_sz;
+
+  char *base = os::reserve_memory(size, false, mtTest);
+  ASSERT_NE(base, (char*) nullptr);
+  ASSERT_TRUE(os::commit_memory(base, size, false));
+
+  for (size_t index = 0; index < pages; index++) {
+    base[index * page_sz] = 'a';
+  }
+
+  os::free_memory_without_uncommit(base, size);
+
+#ifdef __linux__
+  // Check that memory has been freed. Skip on BSD since MADV_DONTNEED doesn't free memory.
+  size_t committed_size;
+  address committed_start;
+  ASSERT_FALSE(os::committed_in_range((address) base, size, committed_start, committed_size));
```

Comment:
Is there a chance of this generating false positives? Do we know if the madvise effect immediate or delayed?

Reply:
That's a good point. Based on the linux [docs](https://man7.org/linux/man-pages/man2/madvise.2.html) it might not happen immediately, causing the test to fail. I'll remove the `committed_in_range check`.  I suppose we could poll with a timeout, but there's still no guarantee the pages actually get freed in a timely manner.